crates/sources/src/sources/uniswap_v3/pool_fetching.rs

```diff
@@ -1,57 +1,237 @@
-use super::graph_api::{RegisteredPools, Token, UniV3SubgraphClient};
-use crate::recent_block_cache::Block;
+use super::graph_api::{PoolData, UniV3SubgraphClient};
 use crate::token_pair::TokenPair;
-use anyhow::Result;
-use ethcontract::{H160, U256};
-use num::BigInt;
+use anyhow::{Context, Result};
+use ethcontract::H160;
+use itertools::{Either, Itertools};
 use reqwest::Client;
-use std::collections::{HashMap, HashSet};
+use std::{
+    collections::{HashMap, HashSet},
+    sync::{Arc, Mutex, Weak},
+    time::{Duration, Instant},
+};
 
 #[async_trait::async_trait]
 pub trait PoolFetching: Send + Sync {
-    async fn fetch(&self, token_pairs: HashSet<TokenPair>, at_block: Block) -> Result<Vec<Pool>>;
+    async fn fetch(&self, token_pairs: &HashSet<TokenPair>) -> Result<Vec<PoolData>>;
 }
 
-pub struct Pool {
-    pub address: H160,
-    pub token0: Token,
-    pub token1: Token,
-    pub sqrt_price: U256,
-    pub liquidity: U256,
-    pub tick: BigInt,
-    pub liquidity_net: HashMap<BigInt, BigInt>,
-    pub fee_tier: U256,
+pub struct CachedPool {
+    pub pool: PoolData,
+    pub updated_at: Instant,
+    pub requested_at: Instant,
 }
 
 pub struct UniswapV3PoolFetcher {
-    pub graph_api: UniV3SubgraphClient,
-    pub registered_pools: RegisteredPools,
+    graph_api: UniV3SubgraphClient,
+    pools_by_token_pair: HashMap<TokenPair, HashSet<H160>>,
+    cache: Mutex<HashMap<H160, CachedPool>>,
+    max_age: Duration,
 }
 
 impl UniswapV3PoolFetcher {
-    pub async fn new(chain_id: u64, client: Client) -> Result<Self> {
+    /// Retrieves all registered pools on Uniswap V3 subgraph, but without `ticks`,
+    /// making the cache values outdated immediately. Cache values are supposed to be updated
+    /// either on fetch or on periodic maintenance update.
+    pub async fn new(chain_id: u64, max_age: Duration, client: Client) -> Result<Self> {
         let graph_api = UniV3SubgraphClient::for_chain(chain_id, client)?;
+        let registered_pools = graph_api.get_registered_pools().await?;
+        tracing::debug!(
+            block = %registered_pools.fetched_block_number, pools = %registered_pools.pools.len(),
+            "initialized registered pools",
+        );
+
+        let mut pools_by_token_pair: HashMap<TokenPair, HashSet<H160>> = HashMap::new();
+        let mut pools: HashMap<H160, CachedPool> = HashMap::new();
+
+        let deep_past = Instant::now()
+            .checked_sub(Duration::from_secs(10000))
+            .context("cant create deep past")?;
```

Comment:
Should this maybe rather use `max_age` in the calculation to make it clearer what the idea is?
And protect against cases where `max_age > 10_000s` (however unlikely).

Reply:
We shouldn't rely on this subtraction working. I'd rather find a different solution like storing Option<Instant>.

Reply:
Tried Option but wasn't satisfied with the code being less clean... I am in favor or putting `max_age` (was even before this comment but forgot to do it since this part of the code was not in constructor and `max_age` was not available there). Why do you think we shouldn't use subtraction?

Reply:
The problem is that the absolute value of an Instant has no meaning so it is possible that the instant starts at 0 when the program starts in which we case the subtraction fails.

Reply:
What's the advantage of initializing the cached pools without the `ticks`? AFAICS `get_pools_and_update_cache()` overwrites the content of the entry anyway (which would also mean its doc comment is not accurrate). So I believe we should be able to skip initializing the cache with invalid data entirely avoiding the `deep_past` thing altogether.

Reply:
> What's the advantage of initializing the cached pools without the `ticks`? AFAICS `get_pools_and_update_cache()` overwrites the content of the entry anyway (which would also mean its doc comment is not accurrate). So I believe we should be able to skip initializing the cache with invalid data entirely avoiding the `deep_past` thing altogether.

The point of initializing the cached pool without `ticks` is that, if we don't do that, the cache is empty, and nothing is updated in "update_recently_used_outdated_pools" unless there are calls to `fn fetch(` that would put some entries in cache and those entries are updated from then. Although I am not sure if this advantage is worth it. I assume everything would work fine without it also...

Reply:
I believe in the interest of trimming the code waiting to fetch data until someone actually wants it is a reasonable trade off. For a long running process and frequently used component like this it should be fine to let the cache build up "naturally".

Reply:
Would it be possible to have the `UniV3SubgraphClient` return a timestamp with the registered pools in order to bootstrap some `deep_past` value? In theory the `updated_at` should be the time of initial indexing from the `UniV3SubgraphClient`.

Reply:
@nlordell Even so, in this moment (contructor initialization) we don't have `ticks` data which makes our cache useless, therefore we don't want to mark it as up to date. I will probably go with the @MartinquaXD suggestion or removing the cache initialization at this point.